src/utils/proxy.js

```diff
@@ -0,0 +1,346 @@
+    const pathHeaderRules = objectForPath(headerRules, requestURL.pathname)
+    if (!isEmpty(pathHeaderRules)) {
+      Object.entries(pathHeaderRules).forEach(([key, val]) => res.setHeader(key, val))
+    }
+    res.writeHead(req.proxyOptions.status || proxyRes.statusCode, proxyRes.headers)
+    proxyRes.on('data', function(data) {
+      res.write(data)
+    })
+    proxyRes.on('end', function() {
+      res.end()
+    })
+  })
+
+  const handlers = {
+    web: (req, res, options) => {
+      const requestURL = new url.URL(req.url, 'http://localhost')
+      req.proxyOptions = options
+      req.alternativePaths = alternativePathsFor(requestURL.pathname).map(p => p + requestURL.search)
+      // Ref: https://nodejs.org/api/net.html#net_socket_remoteaddress
+      req.headers['x-forwarded-for'] = req.connection.remoteAddress || ''
+      return proxy.web(req, res, options)
+    },
+    ws: (req, socket, head) => proxy.ws(req, socket, head),
+  }
+
+  return handlers
+}
+
+async function startProxy(settings = {}, addonUrls, configPath, projectDir) {
+  const functionsServer = settings.functionsPort ? `http://localhost:${settings.functionsPort}` : null
+
+  const proxy = initializeProxy(settings.frameworkPort, settings.dist, projectDir)
+
+  const rewriter = await createRewriter({
+    distDir: settings.dist,
+    jwtRole: settings.jwtRolePath,
+    configPath,
+    projectDir,
+  })
+
+  const server = http.createServer(async function(req, res) {
+    req.originalBody = ['GET', 'OPTIONS', 'HEAD'].includes(req.method) ? null : await createStreamPromise(req, 30)
+
+    if (isFunction(settings.functionsPort, req.url)) {
+      return proxy.web(req, res, { target: functionsServer })
+    }
+    const urlForAddons = addonUrl(addonUrls, req)
+    if (urlForAddons) {
+      return proxy.web(req, res, { target: urlForAddons })
+    }
+
+    rewriter(req, res, match => {
+      const options = {
+        match,
+        addonUrls,
+        target: `http://localhost:${settings.frameworkPort}`,
+        publicFolder: settings.dist,
+        functionsServer,
+        functionsPort: settings.functionsPort,
+        jwtRolePath: settings.jwtRolePath,
+        framework: settings.framework,
+      }
+
+      if (match) return serveRedirect(req, res, proxy, match, options)
+
+      const ct = req.headers['content-type'] ? contentType.parse(req).type : ''
+      if (
+        req.method === 'POST' &&
+        !isInternal(req.url) &&
+        (ct.endsWith('/x-www-form-urlencoded') || ct === 'multipart/form-data')
+      ) {
+        return proxy.web(req, res, { target: functionsServer })
+      }
+
+      proxy.web(req, res, options)
+    })
+  })
+
+  server.on('upgrade', function(req, socket, head) {
+    proxy.ws(req, socket, head)
+  })
+
+  return new Promise(resolve => {
+    server.listen({ port: settings.port }, () => {
+      resolve(`http://localhost:${settings.port}`)
```

Comment:
The original `startProxy` returned an object with `url` and `port` as properties:
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L235
but only `url` was used.
Simplified the function to return just the `url` as a string